File: php/bowling_game/bowling_game.php
<?php

class BowlingGame {
    private $rolls = [];

    // 投球を登録するメソッド
    public function roll($pins) {
        $this->rolls[] = $pins;
    }

    // スコアを計算するメソッド
    public function score() {
        $score = 0;
        $frameIndex = 0;

        for ($frame = 0; $frame < 10; $frame++) {
            if ($this->isStrike($frameIndex)) { // ストライク
                $score += $this->strikeScore($frameIndex);
                $frameIndex++;
            } elseif ($this->isSpare($frameIndex)) { // スペア
                $score += $this->spareScore($frameIndex);
                $frameIndex += 2;
            } else {
                $score += $this->rolls[$frameIndex] + $this->rolls[$frameIndex + 1];
                $frameIndex += 2;
            }
        }

        return $score;
    }

    // ストライクかどうかを判定するメソッド
    private function isStrike($frameIndex) {
        return $this->rolls[$frameIndex] == 10;
    }

    // スペアかどうかを判定するメソッド
    private function isSpare($frameIndex) {
        return $this->rolls[$frameIndex] + $this->rolls[$frameIndex + 1] == 10;
    }

    // ストライクのスコア計算
    private function strikeScore($frameIndex) {
        return 10 + $this->rolls[$frameIndex + 1] + $this->rolls[$frameIndex + 2];
    }

    // スペアのスコア計算
    private function spareScore($frameIndex) {
        return 10 + $this->rolls[$frameIndex + 2];
    }
}

?>

File: lua/bowling_game/bowling_game.lua
-- ボーリングのスコア計算クラス
local BowlingGame = {}

-- 初期化
function BowlingGame:new()
    local game = {
        rolls = {}, -- 倒したピンの数を記録する配列
        MAX_FRAMES = 10 -- 最大フレーム数
    }
    setmetatable(game, self)
    self.__index = self
    return game
end

-- 投球の記録
function BowlingGame:roll(pins)
    table.insert(self.rolls, pins) -- 倒したピンの数を配列に追加
end

-- スコア計算
function BowlingGame:score()
    local score = 0
    local rollIndex = 1 -- 現在の投球のインデックス
    local maxIndex = #self.rolls
    self:roll(0)
    self:roll(0)

    for frame = 1, self.MAX_FRAMES do
        if self:isStrike(rollIndex) then -- ストライクの場合
            score = score + self:strikeScore(rollIndex)
            rollIndex = rollIndex + 1
        elseif self:isSpare(rollIndex) then -- スペアの場合
            score = score + self:spareScore(rollIndex)
            rollIndex = rollIndex + 2
        else -- 通常の投球
            score = score + self.rolls[rollIndex] + self.rolls[rollIndex + 1]
            rollIndex = rollIndex + 2
        end
    end

    return score
end

-- ストライクのボーナス
function BowlingGame:strikeScore(rollIndex)
    return 10 + self.rolls[rollIndex + 1] + self.rolls[rollIndex + 2]
end

-- スペアのボーナス
function BowlingGame:spareScore(rollIndex)
    return 10 + self.rolls[rollIndex + 2]
end

-- ストライクかどうかの判定
function BowlingGame:isStrike(rollIndex)
    return self.rolls[rollIndex] == 10
end

-- スペアかどうかの判定
function BowlingGame:isSpare(rollIndex)
    return self.rolls[rollIndex] + self.rolls[rollIndex + 1] == 10
end

return BowlingGame
